{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Count Snps\n",
    "\n",
    "Code used to count the number of SNPs, differentated between location on exons or introns based on the start and stop coordinates of the gene and exons in Ensembl.\n",
    "Gathers gene, exon, and SNP information from the Ensembl REST API, from the canonical transcript\n",
    "Input: List of official gene symbols\n",
    "Output: Data frame with gene information, exon information, and SNP information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use UCSC human genome browswer to count the number of SNPs across the genome\n",
    "# gather a total, but also seperate by intron and exon \n",
    "\n",
    "import pandas as pd\n",
    "import json \n",
    "import requests\n",
    "import sys\n",
    "import statistics as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in the protein coding genome list\n",
    "genomeFile = open(r\"WGCNA_gene_lists\\20754_human_protein_coding_genes.txt\").read()\n",
    "gene_list_full = genomeFile.split(', ')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_exons(gene_list):\n",
    "    # function for sending dumped data to ensembl api\n",
    "    def fetch_endpoint_POST(server, request, data, content_type):\n",
    "        r = requests.post(server+request,\n",
    "                        headers={ \"Accept\" : content_type},\n",
    "                        data=data )\n",
    "        if not r.ok:\n",
    "            r.raise_for_status()\n",
    "            sys.exit()\n",
    "        # return results\n",
    "        if content_type == 'application/json':\n",
    "            return r.json()\n",
    "        else:\n",
    "            return r.text\n",
    "\n",
    "    # define the server, extension and content type\n",
    "    server = \"http://rest.ensembl.org/\"\n",
    "    con = \"application/json\"\n",
    "    ext_symbols = \"lookup/symbol/homo_sapiens/\"\n",
    "    # create the list of gene symbols\n",
    "    # convert the list into json format\n",
    "    data = json.dumps({ \"symbols\" : gene_list })\n",
    "    # run the query\n",
    "    post_lookup_symbols = fetch_endpoint_POST(server, ext_symbols, data, con)\n",
    "    # Load the data into a pandas dataframe\n",
    "    genes = pd.DataFrame.from_dict(post_lookup_symbols, orient=\"index\")\n",
    "    Ensembl_symbols = genes.loc[:, 'id']\n",
    "    ID_start_end= pd.DataFrame(Ensembl_symbols)\n",
    "    # this generates a dataframe with all of the ensembl IDs for the gene symbols supplied\n",
    "\n",
    "    EnsID = ID_start_end[\"id\"]\n",
    "    \n",
    "    # fetch the canonical transcript ID from each gene to filter exons\n",
    "    canon_t_ids = []\n",
    "    canon_t_starts = []\n",
    "    canon_t_ends = []\n",
    "\n",
    "    for i in EnsID:\n",
    "        ext = \"/overlap/id/\"+i+\"?feature=transcript;species=homo_sapiens\"\n",
    "        r = requests.get(server+ext,headers={ \"Content-Type\" : \"application/json\"})\n",
    "        decoded = r.json()\n",
    "\n",
    "        if not r.ok:\n",
    "            r.raise_for_status()\n",
    "            sys.exit()\n",
    "        transcripts_decoded = r.json()\n",
    "        # format into a dataframe and grab the canonical transcript id\n",
    "        transcripts = pd.DataFrame(transcripts_decoded)\n",
    "\n",
    "        # fetch the start, end, and id of the canonical transcript - append them to the collection lists\n",
    "        canon_t_starts.append(transcripts.loc[(transcripts['is_canonical'] == 1) & (transcripts['Parent'] == i), 'start'].item())\n",
    "        canon_t_ends.append(transcripts.loc[(transcripts['is_canonical'] == 1) & (transcripts['Parent'] == i), 'end'].item())\n",
    "        canon_id = transcripts.loc[(transcripts['is_canonical'] == 1) & (transcripts['Parent'] == i), 'id'].item()\n",
    "        canon_t_ids.append(canon_id)\n",
    "    # assign the accumulator lists as columns of the data frame\n",
    "    ID_start_end['canon transcript id'] = canon_t_ids\n",
    "    ID_start_end['start'] = canon_t_starts\n",
    "    ID_start_end['end'] = canon_t_ends\n",
    "\n",
    "    total_exon_data = []\n",
    "    total_exon_length_list = []\n",
    "    for i in EnsID:\n",
    "        ext = \"/overlap/id/\"+i+\"?feature=exon;species=homo_sapiens\"\n",
    "        r = requests.get(server+ext,headers={ \"Content-Type\" : \"application/json\"})\n",
    "        decoded = r.json()\n",
    "\n",
    "        if not r.ok:\n",
    "            r.raise_for_status()\n",
    "            sys.exit()\n",
    "        decoded = r.json()\n",
    "        exons = pd.DataFrame(decoded)\n",
    "        # filter the exon results for only those that originate from the canonical transcript\n",
    "        exons_canon = exons.loc[exons['Parent'] == ID_start_end.loc[ID_start_end['id'] == i, 'canon transcript id'].item(), :]\n",
    "        exons_canon = exons_canon.reset_index(drop=True)\n",
    "\n",
    "        # turn results into a dictionary with exon id as keys mapping to the start and stop of the exon\n",
    "        exon_dic = {}\n",
    "        total_exon_length = 0\n",
    "        for rowInd in range(len(exons_canon)):\n",
    "            exon_dic[exons_canon.loc[rowInd, 'exon_id']] = (exons_canon.loc[rowInd, 'start'], exons_canon.loc[rowInd, 'end'])\n",
    "            total_exon_length += exons_canon.loc[rowInd, 'end'] - exons_canon.loc[rowInd, 'start']\n",
    "\n",
    "        total_exon_data.append(exon_dic)\n",
    "        total_exon_length_list.append(total_exon_length)\n",
    "        \n",
    "    # add the list of dictionaries as a column of the data frame, and a measure of the length of the gene\n",
    "    ID_start_end['exon start stop'] = total_exon_data\n",
    "    ID_start_end.insert(4, 'gene length', ID_start_end['end'] - ID_start_end['start'])\n",
    "    ID_start_end.insert(6, 'total exon length', total_exon_length_list)\n",
    "\n",
    "    ID_start_end_exon = ID_start_end.reset_index()\n",
    "    ID_start_end_exon = ID_start_end_exon.rename({'index':'gene symbol'}, axis=1)\n",
    "\n",
    "    #print('exons found')\n",
    "    return ID_start_end_exon\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_snps(ens_df):\n",
    "    # iterate over the ensemble id and exon dataframe, grab the SNPs for each gene and clean the dataframe\n",
    "    total_snp_exon_count = []\n",
    "    count_df_list = []\n",
    "\n",
    "    EnsID = ens_df['id']\n",
    "\n",
    "    for i in EnsID:\n",
    "        # get the row number from the exon dataframe\n",
    "        ID_index = ens_df['id'].tolist().index(i)\n",
    "\n",
    "        # get the report of SNPs from the start and stop coordinates of the canonical transcript\n",
    "        server = \"http://rest.ensembl.org/\"\n",
    "        ext = \"/overlap/id/\"+i+\"?feature=variation;species=homo_sapiens\"\n",
    "        r = requests.get(server+ext,headers={ \"Content-Type\" : \"application/json\"})\n",
    "        decoded = r.json()\n",
    "\n",
    "        if not r.ok:\n",
    "            r.raise_for_status()\n",
    "            sys.exit()\n",
    "        decoded = r.json()\n",
    "\n",
    "        snps = pd.DataFrame(decoded)\n",
    "    # ADD THIS CONDITIONAL\n",
    "        if len(snps) > 0:\n",
    "            snps['length'] = snps['end'] - snps['start']\n",
    "            # remove the data that we do not need\n",
    "            snps= snps.drop(labels=['seq_region_name', 'feature_type', 'source', 'strand', 'alleles', 'assembly_name'], axis=1)\n",
    "\n",
    "            levels = list(set(snps['consequence_type']))\n",
    "            # count the number of occurences of each consequence of the SNP - may be able to use for exon/intron differentation at a finer grain\n",
    "            levels_count = {}\n",
    "            for level in levels:\n",
    "                matches = sum(snps['consequence_type'] == level)\n",
    "                levels_count[level] = matches\n",
    "\n",
    "            levels_df = pd.DataFrame(levels_count, index=[i])\n",
    "\n",
    "            count_df_list.append(levels_df)\n",
    "\n",
    "            # accumulators for exon or non exon snps\n",
    "            snp_exon = 0\n",
    "            snp_n_exon = 0\n",
    "            # get the exon range dictionary\n",
    "            exon_dic = ens_df.loc[ID_index, 'exon start stop']\n",
    "            # alternate method: count the snps against the list of exons \n",
    "            for rowInd in range(len(snps)):\n",
    "                in_exon = False\n",
    "                snp_start = snps.loc[rowInd, 'start']\n",
    "                snp_end = snps.loc[rowInd, 'end']\n",
    "                for key in exon_dic.keys():\n",
    "                    # if snp location is inside the exon add and break the loop\n",
    "                    if snp_start >= exon_dic[key][0] and snp_end <= exon_dic[key][1]:\n",
    "                        in_exon = True\n",
    "                        break\n",
    "                # if the count is equal to the length of the keys it means it's in an intron\n",
    "                if in_exon:\n",
    "                    snp_exon += 1\n",
    "                else:\n",
    "                    snp_n_exon += 1\n",
    "\n",
    "\n",
    "        # ADD THIS\n",
    "        else:\n",
    "            snp_exon = None\n",
    "            snp_n_exon = None\n",
    "\n",
    "            temp_df = pd.DataFrame(data = None, columns=['missense_variant'], index=[i])\n",
    "            count_df_list.append(temp_df)\n",
    "        # END ADD SEGMENT\n",
    "            \n",
    "        # append to the accumulator  with ensemble id for binding\n",
    "        total_snp_exon_count.append([i, snp_exon, snp_n_exon,])\n",
    "\n",
    "    count_2_df = pd.DataFrame(data=total_snp_exon_count, columns=['id', 'exon snp', 'not exon snp'])    \n",
    "\n",
    "\n",
    "    total_count_df = pd.concat(count_df_list, axis=0, ignore_index=True)\n",
    "    total_count_df['Total SNPs'] = total_count_df.sum(axis=1, numeric_only=True)\n",
    "\n",
    "    # merge the two data frames together through a temporary column, then remove the column\n",
    "    ens_df['temp'] = list(range(len(ens_df)))\n",
    "    total_count_df['temp'] = list(range(len(total_count_df)))\n",
    "\n",
    "\n",
    "    total_output = ens_df.merge(total_count_df, on='temp', copy=True)\n",
    "    total_output = total_output.drop(labels='temp', axis=1)\n",
    "\n",
    "\n",
    "    total_output2 = total_output.merge(count_2_df, on='id', copy=True)\n",
    "\n",
    "    return total_output2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene symbol</th>\n",
       "      <th>id</th>\n",
       "      <th>canon transcript id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene length</th>\n",
       "      <th>exon start stop</th>\n",
       "      <th>total exon length</th>\n",
       "      <th>splice_donor_region_variant</th>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <th>...</th>\n",
       "      <th>splice_donor_variant</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>inframe_insertion</th>\n",
       "      <th>non_coding_transcript_exon_variant</th>\n",
       "      <th>splice_donor_5th_base_variant</th>\n",
       "      <th>protein_altering_variant</th>\n",
       "      <th>coding_sequence_variant</th>\n",
       "      <th>Total SNPs</th>\n",
       "      <th>exon snp</th>\n",
       "      <th>not exon snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CNNM2</td>\n",
       "      <td>ENSG00000148842</td>\n",
       "      <td>ENST00000369878</td>\n",
       "      <td>102918294</td>\n",
       "      <td>103090222</td>\n",
       "      <td>171928</td>\n",
       "      <td>{'ENSE00001788039': (102918294, 102920101), 'E...</td>\n",
       "      <td>15849</td>\n",
       "      <td>4</td>\n",
       "      <td>5141</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>25</td>\n",
       "      <td>11.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62637.0</td>\n",
       "      <td>6205</td>\n",
       "      <td>56432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CACNA1S</td>\n",
       "      <td>ENSG00000081248</td>\n",
       "      <td>ENST00000362061</td>\n",
       "      <td>201039512</td>\n",
       "      <td>201112426</td>\n",
       "      <td>72914</td>\n",
       "      <td>{'ENSE00004020756': (201112188, 201112426), 'E...</td>\n",
       "      <td>5984</td>\n",
       "      <td>59</td>\n",
       "      <td>178</td>\n",
       "      <td>...</td>\n",
       "      <td>45.0</td>\n",
       "      <td>96</td>\n",
       "      <td>4.0</td>\n",
       "      <td>292.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>28935.0</td>\n",
       "      <td>2949</td>\n",
       "      <td>25986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MYO19</td>\n",
       "      <td>ENSG00000278259</td>\n",
       "      <td>ENST00000614623</td>\n",
       "      <td>36495636</td>\n",
       "      <td>36534868</td>\n",
       "      <td>39232</td>\n",
       "      <td>{'ENSE00003716040': (36534761, 36534868), 'ENS...</td>\n",
       "      <td>3905</td>\n",
       "      <td>43</td>\n",
       "      <td>1331</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>162</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2358.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20355.0</td>\n",
       "      <td>1818</td>\n",
       "      <td>18537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFRP</td>\n",
       "      <td>ENSG00000235718</td>\n",
       "      <td>ENST00000619721</td>\n",
       "      <td>119338942</td>\n",
       "      <td>119346705</td>\n",
       "      <td>7763</td>\n",
       "      <td>{'ENSE00003549514': (119346460, 119346705), 'E...</td>\n",
       "      <td>3934</td>\n",
       "      <td>16</td>\n",
       "      <td>364</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>71</td>\n",
       "      <td>4.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3648.0</td>\n",
       "      <td>1909</td>\n",
       "      <td>1739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ANGPTL6</td>\n",
       "      <td>ENSG00000130812</td>\n",
       "      <td>ENST00000253109</td>\n",
       "      <td>10092338</td>\n",
       "      <td>10102678</td>\n",
       "      <td>10340</td>\n",
       "      <td>{'ENSE00001383912': (10102568, 10102678), 'ENS...</td>\n",
       "      <td>1779</td>\n",
       "      <td>5</td>\n",
       "      <td>271</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>44</td>\n",
       "      <td>2.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4406.0</td>\n",
       "      <td>921</td>\n",
       "      <td>3485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DCLK1</td>\n",
       "      <td>ENSG00000133083</td>\n",
       "      <td>ENST00000360631</td>\n",
       "      <td>35768652</td>\n",
       "      <td>36131382</td>\n",
       "      <td>362730</td>\n",
       "      <td>{'ENSE00003901953': (36131114, 36131382), 'ENS...</td>\n",
       "      <td>8377</td>\n",
       "      <td>14</td>\n",
       "      <td>3723</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "      <td>228.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>129464.0</td>\n",
       "      <td>2904</td>\n",
       "      <td>126560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TVP23C</td>\n",
       "      <td>ENSG00000175106</td>\n",
       "      <td>ENST00000518321</td>\n",
       "      <td>15536980</td>\n",
       "      <td>15563483</td>\n",
       "      <td>26503</td>\n",
       "      <td>{'ENSE00002137998': (15563437, 15563483), 'ENS...</td>\n",
       "      <td>4073</td>\n",
       "      <td>10</td>\n",
       "      <td>1738</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>232.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>21993.0</td>\n",
       "      <td>1492</td>\n",
       "      <td>20501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GPR12</td>\n",
       "      <td>ENSG00000132975</td>\n",
       "      <td>ENST00000405846</td>\n",
       "      <td>26755200</td>\n",
       "      <td>26760786</td>\n",
       "      <td>5586</td>\n",
       "      <td>{'ENSE00001555433': (26760579, 26760786), 'ENS...</td>\n",
       "      <td>4849</td>\n",
       "      <td>1</td>\n",
       "      <td>1243</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2090.0</td>\n",
       "      <td>1766</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ESR1</td>\n",
       "      <td>ENSG00000091831</td>\n",
       "      <td>ENST00000206249</td>\n",
       "      <td>151807682</td>\n",
       "      <td>152103274</td>\n",
       "      <td>295592</td>\n",
       "      <td>{'ENSE00001877305': (151807682, 151808364), 'E...</td>\n",
       "      <td>6319</td>\n",
       "      <td>30</td>\n",
       "      <td>3228</td>\n",
       "      <td>...</td>\n",
       "      <td>22.0</td>\n",
       "      <td>28</td>\n",
       "      <td>3.0</td>\n",
       "      <td>809.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>168255.0</td>\n",
       "      <td>3212</td>\n",
       "      <td>165043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>ENST00000622132</td>\n",
       "      <td>87503017</td>\n",
       "      <td>87600884</td>\n",
       "      <td>97867</td>\n",
       "      <td>{'ENSE00003748396': (87600755, 87600884), 'ENS...</td>\n",
       "      <td>5177</td>\n",
       "      <td>49</td>\n",
       "      <td>462</td>\n",
       "      <td>...</td>\n",
       "      <td>33.0</td>\n",
       "      <td>45</td>\n",
       "      <td>4.0</td>\n",
       "      <td>835.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>76804.0</td>\n",
       "      <td>2260</td>\n",
       "      <td>74544</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene symbol               id canon transcript id      start        end  \\\n",
       "0       CNNM2  ENSG00000148842     ENST00000369878  102918294  103090222   \n",
       "1     CACNA1S  ENSG00000081248     ENST00000362061  201039512  201112426   \n",
       "2       MYO19  ENSG00000278259     ENST00000614623   36495636   36534868   \n",
       "3        MFRP  ENSG00000235718     ENST00000619721  119338942  119346705   \n",
       "4     ANGPTL6  ENSG00000130812     ENST00000253109   10092338   10102678   \n",
       "5       DCLK1  ENSG00000133083     ENST00000360631   35768652   36131382   \n",
       "6      TVP23C  ENSG00000175106     ENST00000518321   15536980   15563483   \n",
       "7       GPR12  ENSG00000132975     ENST00000405846   26755200   26760786   \n",
       "8        ESR1  ENSG00000091831     ENST00000206249  151807682  152103274   \n",
       "9       ABCB1  ENSG00000085563     ENST00000622132   87503017   87600884   \n",
       "\n",
       "   gene length                                    exon start stop  \\\n",
       "0       171928  {'ENSE00001788039': (102918294, 102920101), 'E...   \n",
       "1        72914  {'ENSE00004020756': (201112188, 201112426), 'E...   \n",
       "2        39232  {'ENSE00003716040': (36534761, 36534868), 'ENS...   \n",
       "3         7763  {'ENSE00003549514': (119346460, 119346705), 'E...   \n",
       "4        10340  {'ENSE00001383912': (10102568, 10102678), 'ENS...   \n",
       "5       362730  {'ENSE00003901953': (36131114, 36131382), 'ENS...   \n",
       "6        26503  {'ENSE00002137998': (15563437, 15563483), 'ENS...   \n",
       "7         5586  {'ENSE00001555433': (26760579, 26760786), 'ENS...   \n",
       "8       295592  {'ENSE00001877305': (151807682, 151808364), 'E...   \n",
       "9        97867  {'ENSE00003748396': (87600755, 87600884), 'ENS...   \n",
       "\n",
       "   total exon length  splice_donor_region_variant  3_prime_UTR_variant  ...  \\\n",
       "0              15849                            4                 5141  ...   \n",
       "1               5984                           59                  178  ...   \n",
       "2               3905                           43                 1331  ...   \n",
       "3               3934                           16                  364  ...   \n",
       "4               1779                            5                  271  ...   \n",
       "5               8377                           14                 3723  ...   \n",
       "6               4073                           10                 1738  ...   \n",
       "7               4849                            1                 1243  ...   \n",
       "8               6319                           30                 3228  ...   \n",
       "9               5177                           49                  462  ...   \n",
       "\n",
       "   splice_donor_variant  frameshift_variant  inframe_insertion  \\\n",
       "0                   2.0                  25               11.0   \n",
       "1                  45.0                  96                4.0   \n",
       "2                  42.0                 162                4.0   \n",
       "3                  11.0                  71                4.0   \n",
       "4                   7.0                  44                2.0   \n",
       "5                   4.0                   8                2.0   \n",
       "6                  16.0                  28                NaN   \n",
       "7                   NaN                  15                1.0   \n",
       "8                  22.0                  28                3.0   \n",
       "9                  33.0                  45                4.0   \n",
       "\n",
       "   non_coding_transcript_exon_variant  splice_donor_5th_base_variant  \\\n",
       "0                               105.0                            2.0   \n",
       "1                               292.0                           21.0   \n",
       "2                              2358.0                           17.0   \n",
       "3                                53.0                            9.0   \n",
       "4                                77.0                            5.0   \n",
       "5                               228.0                            2.0   \n",
       "6                               232.0                            2.0   \n",
       "7                                 NaN                            NaN   \n",
       "8                               809.0                           11.0   \n",
       "9                               835.0                           17.0   \n",
       "\n",
       "   protein_altering_variant  coding_sequence_variant  Total SNPs  exon snp  \\\n",
       "0                       1.0                      NaN     62637.0      6205   \n",
       "1                       3.0                      3.0     28935.0      2949   \n",
       "2                       NaN                      1.0     20355.0      1818   \n",
       "3                       NaN                      NaN      3648.0      1909   \n",
       "4                       3.0                      1.0      4406.0       921   \n",
       "5                       3.0                      2.0    129464.0      2904   \n",
       "6                       NaN                      1.0     21993.0      1492   \n",
       "7                       1.0                      NaN      2090.0      1766   \n",
       "8                       NaN                      NaN    168255.0      3212   \n",
       "9                       1.0                      1.0     76804.0      2260   \n",
       "\n",
       "   not exon snp  \n",
       "0         56432  \n",
       "1         25986  \n",
       "2         18537  \n",
       "3          1739  \n",
       "4          3485  \n",
       "5        126560  \n",
       "6         20501  \n",
       "7           324  \n",
       "8        165043  \n",
       "9         74544  \n",
       "\n",
       "[10 rows x 32 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "test_chunk = gene_list_full[50:60]\n",
    "\n",
    "exon_check = get_exons(test_chunk)\n",
    "\n",
    "snp_check = count_snps(exon_check)\n",
    "\n",
    "display(snp_check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function interact with - accepts the list of strings for the genome, a start index, and a stop index\n",
    "def run_gene_list_chunk(gene_list, index_start, index_stop):\n",
    "    # index the list of \n",
    "    gene_chunk = gene_list[index_start:index_stop]\n",
    "    exon_chunk = get_exons(gene_chunk)\n",
    "    snp_results = count_snps(exon_chunk)\n",
    "\n",
    "    return snp_results\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
